### PR TITLE
reproduce and fix the bug in removeSharedRecursive

### PR DESCRIPTION
**Pull request by CheSema**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reproduce the bug described here https://github.com/ClickHouse/ClickHouse/issues/54135

When detached part is deleted, removeSharedRecursive is called for its directory. That call deletes all shared data inside projections despite information about hard linked files inside.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 411a97ff36f9855e4c3f3e85403a9a140ee2ca33 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54430/411a97ff36f9855e4c3f3e85403a9a140ee2ca33/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
</table>

File: src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp [hunk 1]
```diff
@@ -203,6 +203,7 @@ struct RemoveManyObjectStorageOperation final : public IDiskObjectStorageOperati
                 if (unlink_outcome && !keep_all_batch_data && !file_names_remove_metadata_only.contains(fs::path(path).filename()))
                 {
                     objects_to_remove.emplace_back(ObjectsToRemove{std::move(objects), std::move(unlink_outcome)});
+                    removed_files.push_back(path);
```
Comment (tavplubix) [inline src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:206]: The `removed_files` is a bit confusing. Moreover, we already have `remove_paths` in the same scope

**Review by tavplubix — APPROVED** [review]

File: src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp (hunk 1 above)
Comment (CheSema) [reply src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:206]: Ok. Will change this naming.

**Comment by CheSema** [comment]
01051_system_stack_trace
01052_window_view_proc_tumble_to_now
both are flaky

**Pull request merged**